core/src/main/scala/cromwell/core/callcaching/docker/registryv2/flows/HttpFlowWithRetry.scala

```diff
@@ -0,0 +1,120 @@
+  * @tparam T Type of the context data to be passed along with the request
+  * @return A http flow that will retry on request failures and on failed responses as seen fit by the isRetryable method
+  */
+case class HttpFlowWithRetry[T](
+                            httpClientFlow: RetryableHttpFlow[T],
+                            retryBufferSize: Int = 1000,
+                            isRetryable: HttpResponse => Boolean = defaultIsRetryable
+                          ) {
+
+  lazy val flow = GraphDSL.create() { implicit builder =>
+    import GraphDSL.Implicits._
+
+    // Http shape
+    val http = builder.add(httpClientFlow)
+    
+    // Wrap the user context into a pair of (flowContext, httpRequest) to allow for retries
+    val source = builder.add(Flow[(HttpRequest, T)] map {
+      case (httpRequest, flowContext) => (httpRequest, (flowContext, httpRequest))
+    })
+    
+    // Partition Try[HttpResponse] into Success and Failure
+    val partitionResponseTry = builder.add(fanOutTry[HttpResponse, (T, HttpRequest)])
+    
+    val requestSuccessful = partitionResponseTry.out0
+    // Extract the request from the failed try so it can be resubmitted
+    val requestFailed = partitionResponseTry.out1.log("request failed") map toRetryableRequest
+    
+    // Merges requests coming from 3 input ports:
+    // - in(0) takes requests from the outside
+    // - in(1) takes retryable failed requests
+    // - in(2) takes retryable failed responses (with non successful and retryable status code)
+    val mergeRequests = builder.add(Merge[(HttpRequest, (T, HttpRequest))](3))
+    
+    // Splits http response into 3 categories:
+    // - out(0) emits successful responses
+    // - out(1) emits failed but retryable responses
+    // - out(2) emits failed, non-retryable responses
+    val partitionHttpResponse = builder.add(Partition[(HttpResponse, (T, HttpRequest))](3, {
+      // Successful return code
+      case (httpResponse, flowContext) if httpResponse.status.isSuccess() => 0
+      // Failed return code but retryable
+      case (httpResponse, flowContext) if isRetryable(httpResponse) => 1
+      // Failed return code an non retryable
+      case (httpResponse, flowContext) => 2
+    }))
+
+    val responseSuccessful = partitionHttpResponse.out(0) map toTerminalResponse
+    val responseRetryable = partitionHttpResponse.out(1) map toRetryableRequest
+    val responseFailed = partitionHttpResponse.out(2) map toTerminalResponse
+    
+    // Buffer for retryable requests. Will DROP requests when full
+    // Will also delay re submission
+    val retryBuffer = Flow[(HttpRequest, (T, HttpRequest))]
+      .buffer(retryBufferSize, OverflowStrategy.dropHead)
+
+    // Submit outside requests to the first port of the mergeRequests merger 
+    source ~> mergeRequests.in(0)
+    
+    // Submit request to underlying http flow  -  Partition responses: Failure | Success
+    mergeRequests     ~>     http      ~>         partitionResponseTry.in
```

Comment:
It's to try and align the blocks vertically with each other in a meaningful way:
http://doc.akka.io/docs/akka/2.4.16/scala/stream/stream-composition.html#Composing_complex_systems